File: Python/Medium/347-TopKFrequentElements.py
# Description: https://leetcode.com/problems/top-k-frequent-elements/description/

"""
    Approach: Bucket Sort

    Count frequencies of each element using a dictionary (defaultdict).
    Create buckets where index represents frequency (max frequency = n).
    Place each unique element into its corresponding frequency bucket.
    Iterate from highest frequency bucket to collect top k elements.
    Stop when k elements have been collected.
    This avoids sorting and achieves linear time complexity.

    Time complexity: O(n) - Linear frequency counting and bucket collection.
    Space complexity: O(n) - Dictionary and buckets storage.
"""

from typing import List
from collections import defaultdict
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        freq_map = defaultdict(int)
        for num in nums:
            freq_map[num] += 1
        
        buckets = [[] for _ in range(len(nums) + 1)]
        for num, freq in freq_map.items():
            buckets[freq].append(num)
        
        result = []
        for i in range(len(buckets) - 1, -1, -1):
            for num in buckets[i]:
                result.append(num)
                if len(result) == k:
                    return result
        
        return result

# Test
if __name__ == '__main__':
    solution = Solution()
    
    nums1 = [1, 1, 1, 2, 2, 3]
    k1 = 2
    print(f"Input: nums = {nums1}, k = {k1}")
    print(f"Output: {solution.topKFrequent(nums1, k1)}")
    print()
    
    nums2 = [1]
    k2 = 1
    print(f"Input: nums = {nums2}, k = {k2}")
    print(f"Output: {solution.topKFrequent(nums2, k2)}")
    print()
    
    nums3 = [4, 1, -1, 2, -1, 2, 3]
    k3 = 2
    print(f"Input: nums = {nums3}, k = {k3}")
    print(f"Output: {solution.topKFrequent(nums3, k3)}")

File: Java/Medium/347-TopKFrequentElements.java
// Description: https://leetcode.com/problems/top-k-frequent-elements/description/

/*
    Approach: Bucket Sort

    Count frequencies of each element using a HashMap.
    Create buckets where index represents frequency (max frequency = n).
    Place each unique element into its corresponding frequency bucket.
    Iterate from highest frequency bucket to collect top k elements.
    Stop when k elements have been collected.
    This avoids sorting and achieves linear time complexity.

    Time complexity: O(n) - Linear frequency counting and bucket collection.
    Space complexity: O(n) - HashMap and buckets storage.
*/

import java.util.*;
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> freqMap = new HashMap<>();
        for (int num : nums) {
            freqMap.put(num, freqMap.getOrDefault(num, 0) + 1);
        }

        List<Integer>[] buckets = new List[nums.length + 1];
        for (Map.Entry<Integer, Integer> entry : freqMap.entrySet()) {
            int freq = entry.getValue();
            int num = entry.getKey();
            if (buckets[freq] == null) {
                buckets[freq] = new ArrayList<>();
            }
            buckets[freq].add(num);
        }

        List<Integer> result = new ArrayList<>();
        for (int i = buckets.length - 1; i >= 0; i--) {
            if (buckets[i] != null) {
                for (int num : buckets[i]) {
                    result.add(num);
                    if (result.size() == k) {
                        return result.stream().mapToInt(Integer::intValue).toArray();
                    }
                }
            }
        }

        return result.stream().mapToInt(Integer::intValue).toArray();
    }
}

// Test
class Test {
    public static String formatResult(int[] result) {
        int[] sorted = result.clone();
        Arrays.sort(sorted);
        StringBuilder sb = new StringBuilder("[");
        for (int i = 0; i < sorted.length; i++) {
            sb.append(sorted[i]);
            if (i < sorted.length - 1) sb.append(", ");
        }
        sb.append("]");
        return sb.toString();
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // Example 1: nums = [1,1,1,2,2,3], k = 2
        int[] nums1 = {1, 1, 1, 2, 2, 3};
        System.out.println("Output: " + formatResult(solution.topKFrequent(nums1, 2)));

        // Example 2: nums = [1], k = 1
        int[] nums2 = {1};
        System.out.println("Output: " + formatResult(solution.topKFrequent(nums2, 1)));

        // Example 3: nums = [1,2], k = 2
        int[] nums3 = {1, 2};
        System.out.println("Output: " + formatResult(solution.topKFrequent(nums3, 2)));
    }
}
